### ISSUE TITLE
Used seq2seq as autoencoder, first token not predicted right

### ISSUE THREAD

**Issue by exnx** [bentrevett/pytorch-seq2seq#22]
Hi!  Thank you for this repo!  I am using this to create an autoencoder (ie, map the input seq to the same output seq for sentences).  However, the first output token is always predicted as the same thing, the first token in the dictionary.  Has anyone ever run into this problem using this code, or have any tips on how to debug it?  I appreciate it!

**Comment by bentrevett** [comment]
I believe this is because when the tensor to hold all the candidate translations is first created, it is filled with zeros. 

```python
#tensor to store decoder outputs
outputs = torch.zeros(max_len, batch_size, trg_vocab_size).to(self.device)
```

Then when we start putting the candidate translated tokens into the `outputs` tensor, we actually start filling it from index 1, not index 0.

```python
for t in range(1, max_len):
    output, hidden = self.decoder(input, hidden, context)
    outputs[t] = output
```

Hence, the first dimension of `outputs` will always be all zeros. This isn't a problem for calculating the loss as we slice the output tensor, cutting off the first dimension.

```python
output = output[1:].view(-1, output.shape[-1])
```

I imagine your issue is due to not slicing off the first dimension, hence when you use your vocabulary to go from numericalized tensors back to tokens, you'll always have that extra zero which will be the first item in your vocabulary.

**Comment by exnx** [comment]
Ohhhh that is huge. Yes, I did change that part of the code.  I removed the <eos> tag of the output, and removed the <sos> tag of the target.

In that case, I will try just removing the first tag of the output, and leaving the <eos> tag, which I hope should be fine.

Thanks so very much, I was stuck on this for a couple of weeks :(

**Comment by exnx** [comment]
Great it worked!  The first token is now correct.

Minor issue is that the padding is not predicted correctly - instead, the <eos> token is predicted repeatedly at the end.  Though, I can just toss those out anyways.  Just a strange phenomenon that I can't explain.

**Comment by bentrevett** [comment]
Is it the <eos> token predicted repeatedly at the end?

The model will never learn to output a bunch of <pad> tokens after the end of the <eos> token, this is because we set the `ignore_idx` in the loss function to the <pad> token which causes the model to never learn that the <eos> token is followed by a punch of <pad> tokens. 

This is usually fine for inference as you would usually have some decoding loop where you would be checking if the predicted token is <eos>, and if it is then you would just stop decoding - no need to output any <pad> tokens.

**Comment by exnx** [comment]
The <eos> is predicted at the end, yes!  I appreciate you explaining this!  Love it.
